pytket/phir/phirgen_parallel.py

```diff
@@ -0,0 +1,317 @@
+    for group_number in groups:
+        group = groups[group_number]
+        angles2qops: dict[tuple[sympy.Expr | float, ...], dict[str, Any]] = {}
+        for qop in group:
+            if not qop.op.is_gate():
+                append_cmd(qop, ops)
+            else:
+                angles = qop.op.params
+                if tuple(angles) not in angles2qops:
+                    fmt_qop: dict[str, Any] = {
+                        "qop": tket_gate_to_phir[qop.op.type],
+                        "angles": [angles, "pi"],
+                    }
+                    if len(qop.qubits) == 1:
+                        fmt_qop["args"] = [arg_to_bit(qop.qubits[0])]
+                    else:
+                        arg = [arg_to_bit(a) for a in qop.qubits]
+                        fmt_qop["args"] = [arg]
+                    angles2qops[tuple(angles)] = fmt_qop
+                else:
+                    fmt_qop = angles2qops[tuple(angles)]
+                    if len(qop.qubits) == 1:
+                        fmt_qop["args"].append(arg_to_bit(qop.qubits[0]))
+                    else:
+                        arg = [arg_to_bit(a) for a in qop.qubits]
+                        fmt_qop["args"].append(arg)
+        # in the case where the qop is not a gate (a conditional for example)
+        # this branch is skipped because non-gate sub-commands
+        # are always the only member of their group (see process_sub_commands)
+        if len(angles2qops) > 1:
+            pll_block: dict[str, Any] = {"block": "qparallel", "ops": []}
+            for phir_qop in angles2qops.values():
+                pll_block["ops"].append(phir_qop)
+            comment = {"//": f"Parallel {tket_gate_to_phir[qop.op.type]}"}
+            ops.extend((comment, pll_block))
+        else:
+            for phir_qop in angles2qops.values():
+                if len(phir_qop["args"]) > 1:
+                    comment = {"//": "Parallel " + str(qop).split(" q", maxsplit=1)[0]}
+                else:
+                    comment = {"//": str(qop)}
+                ops.extend((comment, phir_qop))
+
+
+def process_shards(
+    shard_layer: ShardLayer, max_parallel_tq_gates: int, max_parallel_sq_gates: int
+) -> dict[int, list[Shard]]:
+    """Break up the shard layer into parallelizable groups."""
+    groups: dict[int, list[Shard]] = {}
+    types2groups: dict[
+        tk.OpType, int
+    ] = {}  # a dict to track the last group in which a gate type was used
+    group_tracker = 0
+    # parallelize shards in the same layer only
+    # keeps a cap on the complexity by not having to look at
+    # the relative ordering of ops in the whole circuit
+    # working within layers guarantees no qubit overlap
+    for shard in shard_layer:
+        # the gate is a quantum op with a fixed number of arguments and
+        # a known machine restriction (i.e. not Barrier, TK2, etc)
+        # add support for barriers later on
```

Comment:
```suggestion
        # TODO(asa): add support for barriers later on
```
Is this the intention?